public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx

```diff
@@ -0,0 +1,368 @@
+} from 'app/types/unified-alerting-dto';
+
+import { stateHistoryApi } from '../../../api/stateHistoryApi';
+import { GRAFANA_DATASOURCE_NAME } from '../../../utils/datasource';
+import { stringifyErrorLike } from '../../../utils/misc';
+import { AlertLabels } from '../../AlertLabels';
+import { CollapseToggle } from '../../CollapseToggle';
+import { STATE_HISTORY_POLLING_INTERVAL } from '../state-history/LokiStateHistory';
+import { LogRecord } from '../state-history/common';
+import { useRuleHistoryRecords } from '../state-history/useRuleHistoryRecords';
+
+const LIMIT_LABELS = 5;
+const LIMIT_EVENTS = 250;
+
+const HistoryEventsList = ({ timeRange }: { timeRange?: TimeRange }) => {
+  const styles = useStyles2(getStyles);
+  const { useGetRuleHistoryQuery } = stateHistoryApi;
+  // Filter state
+  const [eventsFilter, setEventsFilter] = useState('');
+  // form for filter fields
+  const { setValue, register, handleSubmit } = useForm({ defaultValues: { query: '' } }); //  form for search field
+  const from = timeRange?.from.unix();
+  const to = timeRange?.to.unix();
+  const onFilterCleared = useCallback(() => {
+    setEventsFilter('');
+    setValue('query', '');
+  }, [setEventsFilter, setValue]);
+
+  const {
+    currentData: stateHistory,
+    isLoading,
+    isError,
+    error,
+  } = useGetRuleHistoryQuery(
+    {
+      from: from,
+      to: to,
+      limit: LIMIT_EVENTS,
+    },
+    {
+      refetchOnFocus: true,
+      refetchOnReconnect: true,
+      pollingInterval: STATE_HISTORY_POLLING_INTERVAL,
+    }
+  );
+
+  const { historyRecords } = useRuleHistoryRecords(stateHistory, eventsFilter);
+
+  if (isError) {
+    return <HistoryErrorMessage error={error} />;
+  }
+
+  return (
+    <Stack direction="column" gap={1}>
+      <Stack gap={1} direction="row" justifyContent={'space-between'}>
+        <div className={styles.labelsFilter}>
+          <form onSubmit={handleSubmit((data) => setEventsFilter(data.query))}>
+            <SearchFieldInput
+              {...register('query')}
+              showClearFilterSuffix={!!eventsFilter}
+              onClearFilterClick={onFilterCleared}
+            />
+            <input type="submit" hidden />
+          </form>
+        </div>
+      </Stack>
+      <LoadingIndicator visible={isLoading} />
+      <HistoryLogEvents logRecords={historyRecords} />
+    </Stack>
+  );
+};
+
+// todo: this function has been copied from RuleList.v2.tsx, should be moved to a shared location
+const LoadingIndicator = ({ visible = false }) => {
+  const [measureRef, { width }] = useMeasure<HTMLDivElement>();
+  return <div ref={measureRef}>{visible && <LoadingBar width={width} />}</div>;
+};
+
+interface HistoryLogEventsProps {
+  logRecords: LogRecord[];
+}
+function HistoryLogEvents({ logRecords }: HistoryLogEventsProps) {
+  // display log records
+  return (
+    <ul>
+      {logRecords.map((record, index) => {
+        return <EventRow key={index} record={record} />;
```

Comment:
I don't think this is a good key – when refresh is turned on each item might receive a new index and every row will be re-rendered. We should figure out a better way to identify a log line – I think `timestamp + hash(labels)` should be unique enough.